### PR TITLE
Understanding MapReduce in Hadoop

### LINKED ISSUE #752: Understanding MapReduce in Hadoop
### Brief Summary: 
MapReduce is a Hadoop framework that is used for writing applications that can process huge amounts of data. This article provides an understanding of this framework, and how it works. 

### Key Takeaways: 
1. Introduction to MapReduce
2. How MapReduce works
3. The Algorithm used 
4. Phases of the MapReduce model
5. Benefits of MapReduce

### References: 
Please list links to any published content/research that you intend to use to support/guide this article. (If none, please indicate 'N/A'.)

### PR DESCRIPTION
**Pull request by mbaabuones**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.

**NOTE: (1st time contributors - submit ONLY one article to get familiar with the process and style-once published, you can then submit any number of articles.)**
      
#### Grammar Checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical Checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution Guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

This fixes #752 .

### CODE REVIEW

File: articles/understanding-map-reduce-in-hadoop/index.md [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+The following are some of the practical applications of the MapReduce program.
+
+#### E-Commerce
+E-commerce companies such as Walmart, E-Bay, and Amazon use MapReduce to analyze buying behavior. MapReduce provides meaningful information that is used as the basis for developing product recommendations. Some of the information used include site records, e-commerce catalogs, purchase history, and interaction logs. 
+
+#### Social Networks
+The MapReduce programming tool can evaluate certain information on social media platforms such as Facebook, Twitter, and LinkedIn. It can evaluate important information such as who liked your status and who saw your profile. 
+
+#### Entertainment 
+MapReduce is used by Netflix to analyze the clicks and logs of online customers. This information helps the company to suggest movies based on customers’ interests and behavior.
```
Comment (lalith1403) [inline articles/understanding-map-reduce-in-hadoop/index.md:101]: Re-write in active voice: Netflix uses MapReduce to analyze the clicks and logs of online customers.

File: articles/understanding-map-reduce-in-hadoop/index.md [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+[Image Source: A4Academics](http://a4academics.com/images/hadoop/Hadoop-Mapreduce-Architecture.png)
+
+MapReduce architecture consists of various components. A brief description of these components can improve our understanding of how MapReduce works. 
+- **Job:** This is an execution of a reducer or mapper across a dataset.
+- **Task:** This is an execution of a reducer or mapper on a section of data
+- **Job Tracker:** This tracker plays the role of scheduling jobs and tracking all jobs assigned to the task tracker.
+- **Task Tracker:** This tracker plays the role of tracking tasks and reporting the status of tasks to the job tracker.  
+- **Input data:** This is the data used for processing in the mapping phase.
+- **Output data:** This is the result of mapping and reducing.
+- **Client:** The MapReduce receives jobs from clients. It can accept jobs from many clients.
```
Comment (lalith1403) [inline articles/understanding-map-reduce-in-hadoop/index.md:31]: What is a client? Is it an application software? A server? Emphasize what it is and then its utility.

File: articles/understanding-map-reduce-in-hadoop/index.md [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+#### MapReduce Architecture
+The following diagram shows a MapReduce architecture.
+
+![MapReduce Architecture](/engineering-education/understanding-mapreduce-in-hadoop/map-reduce-architecture.png)
+
+[Image Source: A4Academics](http://a4academics.com/images/hadoop/Hadoop-Mapreduce-Architecture.png)
+
+MapReduce architecture consists of various components. A brief description of these components can improve our understanding of how MapReduce works. 
+- **Job:** This is an execution of a reducer or mapper across a dataset.
+- **Task:** This is an execution of a reducer or mapper on a section of data
```
Comment (lalith1403) [inline articles/understanding-map-reduce-in-hadoop/index.md:26]: Why is this distinction so important? The difference between a Job and a Task? If the only difference is job is for the entire dataset and the task is for a part of the dataset, why would it be named this way? What is the need? Elaborate on that!

File: articles/understanding-map-reduce-in-hadoop/index.md [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+![Job Trackers and Task Trackers](/engineering-education/understanding-mapreduce-in-hadoop/job-trackers-and-task-trackers.png)
+
+[Image Source: CNBlogs](https://images2015.cnblogs.com/blog/222972/201512/222972-20151215101500256-2116703561.png)
+
+#### Phases of MapReduce
+The MapReduce program is executed in three main phases: mapping, shuffling, and reducing. There is also an optional phase known as the combiner phase.
+
+##### Mapping Phase
+This is the first phase of the program. There are two steps in this phase: splitting and mapping. In this phase, a dataset is split into equal units called chunks. These chunks can be termed as input splits. MapReduce uses the input splits as inputs. They are composed of two parts: keys and values. In the mapping step, the required action is performed using these subsets. This step contains a coding logic that is applied to a specific number of data blocks. The output of this phase is in the form of keys and values (k, v).
```
Comment (lalith1403) [inline articles/understanding-map-reduce-in-hadoop/index.md:58]: equal units called chunks->equal units called _chunks_

Comment (lalith1403) [inline articles/understanding-map-reduce-in-hadoop/index.md:58]: What are you referring to when you say they are composed of two parts: keys and values? What is 'they'? Also, after they are composed of two parts, this para is a bit difficult to comprehend. Can you use the previous example that I have outlined and make the changes accordingly?

**Review by lalith1403 — CHANGES_REQUESTED** [review]
Well structured article. I understood the concept of MapReduce really well. A few places I had to pay extra attention to understand. Such places I have asked you to reword to enhance readability. Also, whenever you introduce new keywords such as map, reduce etc, you may use italics or bold to highlight them. This way the attention of the reader is captured.

**Comment by hectorkambow** [comment]
plagiarism check done

File: articles/understanding-map-reduce-in-hadoop/index.md [hunk 5]
```diff
@@ -0,0 +1,118 @@
+MapReduce is a component of the Apache Hadoop ecosystem, a framework that enhances massive data processing. Other components of Apache Hadoop include Hadoop Distributed File System (HDFS), Yarn, and Apache Pig.
```
[the remark this answers is not in the document]
Comment (mbaabuones) [reply articles/understanding-map-reduce-in-hadoop/index.md]: Done.

File: articles/understanding-map-reduce-in-hadoop/index.md [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+This article provides an understanding of MapReduce in Hadoop. It will enable readers to gain insights on how vast volumes of data is simplified and how MapReduce is used in real-life applications. 
+
+### Introduction to MapReduce in Hadoop
+MapReduce is a Hadoop framework for writing applications that can process vast amounts of data on large clusters. It can also be termed a programming model in which large datasets can be processed across computer clusters. This application allows data to be stored in a distributed form. It simplifies huge volumes of data and large scale computing. 
+
+There are two main tasks in MapReduce: map and reduce. The map task is done first, followed by the reduce task. In the map job, the input dataset is split into chunks. Map task processes these chunks in a parallel manner. The map outputs are used as inputs for the Reduce tasks. Reducers process the intermediate data from the maps into smaller tuples, which reduce the tasks. This is the final output of the framework.
```
[the remark this answers is not in the document]
Comment (mbaabuones) [reply articles/understanding-map-reduce-in-hadoop/index.md]: Noted. I have incorporated them in the article. Thanks.

File: articles/understanding-map-reduce-in-hadoop/index.md [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+There are two main tasks in MapReduce: map and reduce. The map task is done first, followed by the reduce task. In the map job, the input dataset is split into chunks. Map task processes these chunks in a parallel manner. The map outputs are used as inputs for the Reduce tasks. Reducers process the intermediate data from the maps into smaller tuples, which reduce the tasks. This is the final output of the framework.
+
+The MapReduce framework enhances the scheduling and monitoring of tasks. The failed tasks are re-executed by the framework. This framework can be used easily, even by programmers with little expertise in distributed processing. MapReduce can be implemented using various programming languages such as Java, Hive, Pig, Scala, and Python. 	
+ 
+### How MapReduce in Hadoop works
+An overview of MapReduce Architecture and MapReduce's phases will help us understand how MapReduce in Hadoop works.
```
[the remark this answers is not in the document]
Comment (mbaabuones) [reply articles/understanding-map-reduce-in-hadoop/index.md]: Done.

File: articles/understanding-map-reduce-in-hadoop/index.md (hunk 1 above)
Comment (mbaabuones) [reply articles/understanding-map-reduce-in-hadoop/index.md]: Done.

File: articles/understanding-map-reduce-in-hadoop/index.md (hunk 2 above)
Comment (mbaabuones) [reply articles/understanding-map-reduce-in-hadoop/index.md]: I have explained what a client is.

File: articles/understanding-map-reduce-in-hadoop/index.md (hunk 3 above)
Comment (mbaabuones) [reply articles/understanding-map-reduce-in-hadoop/index.md]: I have revised accordingly.

File: articles/understanding-map-reduce-in-hadoop/index.md (hunk 4 above)
Comment (mbaabuones) [reply articles/understanding-map-reduce-in-hadoop/index.md]: Done.

Comment (mbaabuones) [reply articles/understanding-map-reduce-in-hadoop/index.md]: I have revised this section.

**Comment by mbaabuones** [comment]
@lalith1403 I have finished working on the changes.

**Pull request merged**